File: backend/tiggle-root/tiggle/src/main/java/com/side/tiggle/domain/category/api/CategoryApiController.java
package com.side.tiggle.domain.category.api;

import com.side.tiggle.domain.category.dto.CategoryDto;
import com.side.tiggle.domain.category.dto.controller.CategoryUpdateReqDto;
import com.side.tiggle.domain.category.dto.resp.CategoryRespDto;
import com.side.tiggle.domain.category.model.CategoryType;
import com.side.tiggle.domain.category.service.CategoryService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.stream.Collectors;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/v1/category")
public class CategoryApiController {

    private final CategoryService categoryService;
    
    @PostMapping
    public ResponseEntity<CategoryRespDto> createCategory(@RequestBody CategoryDto categoryDto) {
        return new ResponseEntity<>(CategoryRespDto.fromEntity(categoryService.createCategory(categoryDto)), HttpStatus.OK);
    }
    

    @GetMapping("/{id}")
    public ResponseEntity<CategoryRespDto> getCategory(@PathVariable("id") Long categoryId) {
        return new ResponseEntity<>(CategoryRespDto.fromEntity(categoryService.getCategory(categoryId)), HttpStatus.OK);
    }

    @GetMapping("/type/{type}")
    public ResponseEntity<List<CategoryRespDto>> getCategory(@PathVariable("type") CategoryType type) {
        return new ResponseEntity<>(categoryService.getCategoryType(type).stream()
                .map(CategoryRespDto::fromEntity)
                .collect(Collectors.toList()), HttpStatus.OK);
    }

    @GetMapping("/all")
    public ResponseEntity<List<CategoryRespDto>> getAllCategory() {
        return new ResponseEntity<>(categoryService.getAllCategory()
                .stream()
                .map(CategoryRespDto::fromEntity)
                .collect(Collectors.toList()), HttpStatus.OK);
    }
    
    @PutMapping("/{id}")
    public ResponseEntity<CategoryRespDto> updateCategory(
            @PathVariable("id") Long categoryId,
            @RequestBody CategoryUpdateReqDto dto
    ) {
        return new ResponseEntity<>(CategoryRespDto.fromEntity(categoryService.updateCategory(categoryId, dto)), HttpStatus.OK);
    }
    
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteCategory(
            @PathVariable("id") Long categoryId
    ) {
        categoryService.deleteCategory(categoryId);
        return new ResponseEntity<>(null, HttpStatus.OK);
    }
}

File: backend/tiggle-root/tiggle/src/main/kotlin/com/side/tiggle/domain/category/api/CategoryApiController.kt
package com.side.tiggle.domain.category.api

import com.side.tiggle.domain.category.dto.req.CategoryCreateReqDto
import com.side.tiggle.domain.category.dto.req.CategoryUpdateReqDto
import com.side.tiggle.domain.category.dto.resp.CategoryListRespDto
import com.side.tiggle.domain.category.service.CategoryService
import com.side.tiggle.global.common.ApiResponse
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import com.side.tiggle.global.common.constants.HttpHeaders
import jakarta.validation.Valid
import jakarta.validation.constraints.Min
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*

@Validated
@RestController
@RequestMapping("/api/v1/category")
class CategoryApiController(
    private val categoryService: CategoryService
) {

    @PostMapping
    fun createCategory(
        @RequestBody @Valid categoryCreateReqDto: CategoryCreateReqDto,
        @RequestHeader(name = HttpHeaders.MEMBER_ID) memberId: Long
    ): ResponseEntity<ApiResponse<Nothing>> {
        categoryService.createCategory(categoryCreateReqDto, memberId)
        return ResponseEntity
            .status(HttpStatus.CREATED)
            .body(ApiResponse.success(null, message = "카테고리가 생성되었습니다."))
    }

    @GetMapping()
    fun getCategoryByMemberIdOrDefaults(
        @RequestHeader(name = HttpHeaders.MEMBER_ID) memberId: Long
    ): ResponseEntity<ApiResponse<CategoryListRespDto>> {
        val categories = categoryService.getCategoryByMemberIdOrDefaults(memberId)
        return ResponseEntity
            .ok(ApiResponse.success(categories))
    }

    @PutMapping("/{id}")
    fun updateCategory(
        @PathVariable("id") @Min(1) categoryId: Long,
        @RequestBody @Valid dto: CategoryUpdateReqDto
    ): ResponseEntity<ApiResponse<Nothing>> {
        categoryService.updateCategory(categoryId, dto)
        return ResponseEntity
            .ok(ApiResponse.success(null, message = "카테고리가 수정되었습니다."))
    }

    @DeleteMapping("/{id}")
    fun deleteCategory(
        @PathVariable("id") @Min(1) categoryId: Long
    ): ResponseEntity<ApiResponse<Nothing>> {
        categoryService.deleteCategory(categoryId)
        return ResponseEntity
            .ok(ApiResponse.success(null, message = "카테고리가 삭제되었습니다."))
    }
}
